### PR TITLE
Remove `AsyncIterator: Sendable` requirement from merge

### PR DESCRIPTION
**Pull request by FranzBusch**
# Motivation
Currently a lot of the operator implementations in here that consume other `AsyncSequence`s require the `AsyncIterator` to be `Sendable`. This is mostly due to the fact that we are calling `makeAsyncIterator` on the upstream `AsyncSequence` and then pass that iterator around to various newly spawned `Task`s. This has two downsides:
1. It only allows users to use operators like `merge` if their `AsyncSequence.AsyncIterator` is `Sendable`
2. In merge we are creating new `Task`s for every new demand. Creating `Task`s is not cheap.

My main goal of this PR was to remove the `Sendable` constraint from `merge`.

# Modification
This PR overhauls the complete inner workings of the `AsyncMerge2Sequence`. It does a couple of things:
1. The main change is that instead of creating new `Task`s for every demand, we are creating one `Task` when the `AsyncIterator` is created. This task has as child task for every upstream sequence.
2. When calling `next` we are signalling the child tasks to demand from the upstream
3. A new state machine that is synchronizing the various concurrent operations that can happen
4. Handling cancellation since we are creating a bunch of continuations.

# Result
In the end, this PR swaps the implementation of `AsyncMerge2Sequence` and drops the `Sendable` constraint and passes all tests. Furthermore, on my local performance testing I saw up 50% speed increase in throughput.

# Open points
1. I need to make this sequence re-throwing but before going down that rabbit whole I wanna get buy-in on the implementation.
2. We should discuss and document if `merge` and other operators are hot or cold, i.e. if they only request if they got downstream demand
3. I need to switch `AsyncMerge3Sequence` over to the same iplementation

### CODE REVIEW

File: Package.swift
Comment (FranzBusch) [inline Package.swift:20]: I was unsure if we should add this dependency. It makes the implementation a lot better and IMO is a lower level package. Open to discuss this :)

**Comment by FranzBusch** [comment]
Switching this from `CheckedContinuation` to `UncheckedContinuation` is improving the performance by a lot:

Current main: 1x baseline
This PR with Checked: 1.5x
This PR with Unchecked: 3.5x

**Comment by FranzBusch** [comment]
@phausler Yeah no rush on this! Enjoy your vacation and we can talk afterwards. I just wanted to put this up for discussion. I agree that re throwing is kinda coming the way but I am seeing up to 3.5 times performance improvements with this. However, performance is no the main driver, it really is the `Sendable` annotation. Let's discuss this once you are back!

File: Sources/AsyncAlgorithms/Merge2StateMachine.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/Merge2StateMachine.swift:38]: do you think there is a way we can approach this without `Task.select`? That is the other perf bottle neck (which could be either solved by not needing it, or solved by improving it at a language level)

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:166]: since this is being validated by tests we can use the non checked variant

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:553]: the preconditions in here make my spidey senses tingle.... any fatal error will become a bug (in my experience). is it possible to somehow refactor these to not be?

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:602]: it might be good to (if we decide these are useful to keep) list the state that is invalid and unify that with the same verbiage as the other fatal error cases.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:612]: stylistic nit; most folks have wide enough monitors to see the comment, so no need to limit code to 40 col... (cause if you are editing swift on a non emulated VT100... well time to upgrade :) )

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:694]: this feels like it somehow could not be fully true... do we have tests that validate this?

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:727]: small nit: we rarely use past tense for names of cases. `.upstreamFailure` might be a better name.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:751]: some of these long names and numerous associated types kinda almost infer structness of the associations. Would it be better to change them to structures instead of effectively tuples?

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [inline Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:772]: might be worth it to abstract this to a static inline variable which ought to have the same perf as this. but then give it a name.

**Review by phausler — APPROVED** [review]
Overall this looks good, it could use some sprucing up with regards to the locks before merging (id rather not introduce a new locking scheme and keep to the managed critical state stuff if at all possible.

It took me a bit to follow all of the logic but it seems relatively sound (excluding the whole fatal error cases marked in the review).

With changing to the unchecked variants I think we can eek out some more perf!

**Comment by phausler** [comment]
One required item for merging this: it needs to definitely handle rethrowing cases (and we should validate that is true)

File: Sources/AsyncAlgorithms/Merge2StateMachine.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/Merge2StateMachine.swift:96]: This is the old code. I just copied this since other algos were using this and I did not want to change them.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:122]: I tried to adopt this but `ManagedCriticalState` does only support using a lock with a closure. For the `next` method I would like to manually lock and unlock to be able to hold the lock across the `withUnsafeContinuation` method.

File: Sources/AsyncAlgorithms/Merge2StateMachine.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/Merge2StateMachine.swift:38]: Same here, not my code just copied over.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:553]: IMO, we should keep these preconditions. The state that I am preconditioning in should never been hit and I haven't been able to force the state machine into these states with my tests. If we ever hit one of those, I consider this a good thing because the assumptions I have made are wrong.

The good thing is this won't become a subtle bug but rather a crash. We have been using this in NIO for a long time and it helped us flush out bugs in all our state machines.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:612]: I am using newlines here mostly to make reading it easier not because of screen width. I would prefer to keep that to maintain readability.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:694]: Various tests are validating that indirectly. For this precondition to fail we would need to get into a state where we have a suspended call to `next()` and things in the buffer. However, when we call `next` and have things in the buffer we are returning them right away. We are only suspending if the buffer is empty. Furthermore, since we are holding the lock across the `withUnsafeContinuation` call in `next` nothing can get in-between us checking if the buffer is empty and us suspending.

When any of the upstreams produces a new element, it will acquire the lock and then we come into this case. If we have a continuation we are guaranteed to have an empty buffer since we would have never suspended in the first place if we had something in the buffer.

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:751]: In general I am with you; however, this would introduce just a struct for one case of an action. My personal opinion, is that this is overkill for something internal. If you this is important to you, I am happy to change it though. Your call :)

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift:772]: Agreed, I also wanna change this around since I want to be able to support the merge2 and merge3 algorithm with the same state machine.

**Comment by FranzBusch** [comment]
@phausler I just updated the PR again. To summarise what changed from your initial review:
- My implementation supports rethrowing now
- I split up the implementation into separate files to aid with readability
- I swapped `merge3` to also make use of the new state machine
- I renamed the internal `upstreamThrew` case to `upstreamFailure`
- I added more tests validating various edge cases
- Aligned precondition messages

I replied to some of your feedback inline but I haven't changed the following:
- Use `ManagedCriticalState`: I can't adopt this since I need to be able to manually `lock` and `unlock` in the case when I need to suspend on `next()`
- Remove preconditions: I went through the code again and checked these preconditions and I wasn't able to come up with a test to break them. There might still be an edge case I haven't thought about, but I would rather let this crash then to transition into an unknown state.
- Line breaks: I find them easier to read but if you want me to use a single line I am happy to change (maybe it would be good to get swift format setup in this repo?)

File: Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift
Comment (phausler) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift]: Holding a lock across a suspension is almost always a really bad idea; it can cause some pretty heinous fall out.

Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift]: Holding a lock across `with*Continuation` is safe to do since the closure is executed right away. We have been doing this in various places of the NIO ecosystem now.

Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift]: To further clarify this is not holding a lock across a suspension point but just across the `with*Continuation` call.

Comment (phausler) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift]: so there is only 1 way to do that safely (and it is a very edge case scenario).

```
lock()
await withUnsafeContinuation { continuation in
  self.continuation = continuation
  unlock()
}
```

It is very easy to modify that pattern into some really gnarly failure modes. From a maintainability standpoint id rather make sure there is no other way of doing that.

Comment (FranzBusch) [reply Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift]: Right I am using exactly that pattern just with calls to the state inbetween. I agree with you that this needs to be done with care.
However, if don’t hold the lock across that call, the state handling becomes trickier since it introduces a potential interleaving between next and closure of withUnsafeContinuation where everything else can happen

**Comment by phausler** [comment]
@swift-ci please test

**Comment by FranzBusch** [comment]
@phausler Just rebased this PR. Feel free to merge it if you are happy with it!

**Pull request merged**
